### PR TITLE
Fix yii\rest\Serializer serialize ArrayDataProvider with pagination bug

### PR DESCRIPTION
**Pull request by dcb9**
| Q | A |
| --- | --- |
| Is bugfix? | yes |
| New feature? | no |
| Breaks BC? | yes/no |
| Tests pass? | yes |
| Fixed issues | #11977 |

```
➜  yii2 git:(fixRestSerializeArrayDataProvider) ./vendor/bin/phpunit --group=rest
PHPUnit 4.8.24-15-g3c4becb by Sebastian Bergmann and contributors.

....

Time: 895 ms, Memory: 46.50Mb

OK (4 tests, 39 assertions)
```

### CODE REVIEW

File: framework/rest/Serializer.php [hunk 1]
```diff
@@ -171,7 +172,11 @@ protected function getRequestedFields()
      */
     protected function serializeDataProvider($dataProvider)
     {
-        $models = $this->serializeModels($dataProvider->getModels());
+        $models = $dataProvider->getModels();
+        if ($dataProvider instanceof ArrayDataProvider) {
```
Comment (AnatolyRugalev) [inline framework/rest/Serializer.php]: Do we really need this check? I think this bug may appear in any kind of `DataProvider`, not only `ArrayDataProvider`

**Comment by klimov-paul** [comment]
Resolved by commit cd8c01f555e425d4654775002ae5f44152e0c4b1

**Comment by dcb9** [comment]
thanks @klimov-paul

**Comment by LAV45** [comment]
@klimov-paul , `array_values()` removes the key which has been added using `->indexBy('tag')`

```php
class ManufacturerController extends \yii\rest\Controller
{
    public $serializer = [
        'class' => 'yii\rest\Serializer',
        'collectionEnvelope' => 'manufacturers',
    ];

    public function actionIndex()
    {
        $query = Manufacturer::find()
            ->indexBy('tag');

        return new ActiveDataProvider([
            'query' => $query,
        ]);
    }
}
```

**Comment by klimov-paul** [comment]
> removes the key which has been added using ->indexBy('tag')

I can't see a problem here: data provider serialization is meant to return array of objects in JSON format, so it does it now.
If this behavior is not desiarable you should extend `Serializer` and implement `serializeDataProvider()` in your own way.

**Comment by cebe** [comment]
still this change is a BC break.

**Comment by klimov-paul** [comment]
> still this change is a BC break.

Why? Using of named keys is not something documented for the DataProvider serialization. Serializing of the list of rows expects production of the **list** of equal entities, not a data structure, where each row has its key and thus is logic meaning.

**Comment by LAV45** [comment]
This API design was given to me as inheritance)
before change
```js
{
    code: 0,
    text: "ok",
    result: {
        manufacturers: {
            its_soft: {
                id: 1,
                name: '...',
                tag: 'its_soft'
            },
            tehnokom: {
                id: 2,
                name: '...',
                tag: 'tehnokom'
            }
        }
    }
}
```
after change
```js
{
    code: 0,
    text: "ok",
    result: {
        manufacturers: [
            {
                id: 1,
                name: '...',
                tag: 'its_soft'
            },
            {
                id: 2,
                name: '...',
                tag: 'tehnokom'
            }
        ]
    }
}
```

**Comment by klimov-paul** [comment]
As I said this is not meant to be working in this way. If it should - the whole change should be reverted and this issue closed with resolution: 'not bug but feature'.

You can still return the list of models without data provider wrapped around them - in this case keys will be preserved.

**Comment by AnatolyRugalev** [comment]
@LAV45 you may implement your own `Serializer` in this case. I'm agree with @klimov-paul, this is not documented feature (not a feature at all) and `DataProvider`  is not designed to provide structures except **list**.

Also I agree with @cebe, this fix breaks BC.

Original issue was about REST pagination problems of ArrayDataProvider, which may be resolved by changing 4th parameter of `array_slice` to `false` [here](https://github.com/yiisoft/yii2/blob/217d50e4ff690d651d910e61cb3601044031e185/framework/data/ArrayDataProvider.php#L91)

**Comment by cebe** [comment]
@klimov-paul even though it was not documented, the current behavior with models that have non-integer keys was to return a JSON object indexed by the $key property of the dataprovider.
DataProvider itself was designed to work with keys, we even have separation of data and key values in it. It is not correct to assume the expectation of array in this case but a fix like proposed in #12683 should be appliced to make it configurable. It is not possible to autodetect whether indexed output is wanted or not, so a config option should be added.

**Comment by klimov-paul** [comment]
> even though it was not documented, the current behavior with models that have non-integer keys was to return a JSON object indexed by the $key property of the dataprovider.

This behavior is inconsistent - this entire issue was about this.
While `ActiveDataProvider` iterates over pages with native keys per each page, `ArrayDataProvider` preserves orginal keys. Thus, in general case, for `ActiveDataProvider` each page will be converted to JSON array, while for `ArrayDataProvider` first page will be an array, while second and so on becomes object!

For me `DataProviver::getModels()` should return the list of records, e.g. array in JSON representation - not the data strucure - object in JSON representation.

> DataProvider itself was designed to work with keys, we even have separation of data and key values in it

It does work with keys but in different manner: there is a method `DataProvider::getKeys()`, which return  s keys for the data. `ActiveDataProvider` uses model primaky key for it, while `ArrayDataProvider` relies on native source array keys. That is why `ArrayDataProvider` can not strip them while splitting data into pages.

As I already said: in order to return a data-structure you can return array of models instead of `DataProvider`. Being indexed by cutom keys it will be serialized correctly.

> fix like proposed in #12683 should be appliced to make it configurable. It is not possible to autodetect whether indexed output is wanted or not, so a config option should be added.

Let it be. However such option should be disabled by default.

**Comment by cebe** [comment]
fixed by  	04848a6

**Pull request closed without merging**